File: backend/java/bcsms-api/src/main/java/com/bursa/bcsms/domain/entity/ServiceRequest.java
package com.bursa.bcsms.domain.entity;

import com.bursa.bcsms.common.exception.DomainException;
import com.bursa.bcsms.domain.enums.Priority;
import com.bursa.bcsms.domain.enums.RequestStatus;
import com.bursa.bcsms.domain.valueobject.Location;
import jakarta.persistence.*;

import java.time.Instant;
import java.util.*;

@Entity
@Table(name = "service_requests")
public class ServiceRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "title", nullable = false, length = 200)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "category_id", nullable = false)
    private UUID categoryId;

    @Column(name = "citizen_id", nullable = false)
    private UUID citizenId;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    private RequestStatus status;

    @Enumerated(EnumType.STRING)
    @Column(name = "priority", length = 50)
    private Priority priority;

    @Column(name = "assigned_department_id")
    private UUID assignedDepartmentId;

    @Column(name = "assigned_employee_id")
    private UUID assignedEmployeeId;

    @Embedded
    private Location location;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @OneToMany(mappedBy = "serviceRequest", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<StatusHistoryEntry> statusHistory = new ArrayList<>();

    @OneToMany(mappedBy = "serviceRequest", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<Comment> comments = new ArrayList<>();

    @OneToMany(mappedBy = "serviceRequest", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<Attachment> attachments = new ArrayList<>();

    protected ServiceRequest() {
    }

    public ServiceRequest(UUID id, String title, UUID categoryId, UUID citizenId,
                          Instant createdAt, String description, Location location) {
        if (title == null || title.isBlank()) {
            throw new DomainException("Service request title is required.");
        }
        if (categoryId == null) {
            throw new DomainException("Category ID is required.");
        }
        if (citizenId == null) {
            throw new DomainException("Citizen ID is required.");
        }

        this.id = id != null ? id : UUID.randomUUID();
        this.title = title.trim();
        this.categoryId = categoryId;
        this.citizenId = citizenId;
        this.description = description != null ? description.trim() : null;
        this.location = location;
        this.status = RequestStatus.NEW;
        this.createdAt = createdAt != null ? createdAt : Instant.now();
    }

    // ──────────────────────────────────────────────
    // State Machine Lifecycle Transitions
    // ──────────────────────────────────────────────

    /**
     * Transitions: NEW -> REVIEWING
     */
    public void startReview(UUID changedByUserId, Instant utcNow) {
        ensureValidTransition(RequestStatus.NEW, RequestStatus.REVIEWING);
        transitionTo(RequestStatus.REVIEWING, changedByUserId, utcNow, null);
    }

    /**
     * Transitions: REVIEWING -> ASSIGNED
     */
    public void assign(UUID departmentId, UUID employeeId, Priority priority, UUID changedByUserId, Instant utcNow) {
        ensureValidTransition(RequestStatus.REVIEWING, RequestStatus.ASSIGNED);

        if (departmentId == null) {
            throw new DomainException("Department ID is required for assignment.");
        }
        if (priority == null) {
            throw new DomainException("Priority is required for assignment.");
        }

        this.assignedDepartmentId = departmentId;
        this.assignedEmployeeId = employeeId;
        this.priority = priority;
        transitionTo(RequestStatus.ASSIGNED, changedByUserId, utcNow, null);
    }

    /**
     * Transitions: ASSIGNED -> IN_PROGRESS
     */
    public void startProgress(UUID changedByUserId, Instant utcNow) {
        ensureValidTransition(RequestStatus.ASSIGNED, RequestStatus.IN_PROGRESS);
        transitionTo(RequestStatus.IN_PROGRESS, changedByUserId, utcNow, null);
    }

    /**
     * Transitions: IN_PROGRESS -> RESOLVED
     */
    public void resolve(UUID changedByUserId, Instant utcNow, String note) {
        ensureValidTransition(RequestStatus.IN_PROGRESS, RequestStatus.RESOLVED);
        transitionTo(RequestStatus.RESOLVED, changedByUserId, utcNow, note);
    }

    /**
     * Transitions: RESOLVED -> CLOSED
     */
    public void close(UUID changedByUserId, Instant utcNow, String note) {
        ensureValidTransition(RequestStatus.RESOLVED, RequestStatus.CLOSED);
        transitionTo(RequestStatus.CLOSED, changedByUserId, utcNow, note);
    }

    /**
     * Transitions: RESOLVED -> IN_PROGRESS
     */
    public void reopen(UUID changedByUserId, Instant utcNow, String note) {
        ensureValidTransition(RequestStatus.RESOLVED, RequestStatus.IN_PROGRESS);
        transitionTo(RequestStatus.IN_PROGRESS, changedByUserId, utcNow, note);
    }

    /**
     * Transitions: NEW or REVIEWING -> REJECTED
     */
    public void reject(UUID changedByUserId, Instant utcNow, String note) {
        if (this.status != RequestStatus.NEW && this.status != RequestStatus.REVIEWING) {
            throw new DomainException("Cannot reject a request in '" + this.status + "' status. Only 'NEW' or 'REVIEWING' requests can be rejected.");
        }
        transitionTo(RequestStatus.REJECTED, changedByUserId, utcNow, note);
    }

    /**
     * Transitions: NEW, REVIEWING, or ASSIGNED -> CANCELLED
     */
    public void cancel(UUID changedByUserId, Instant utcNow) {
        if (this.status != RequestStatus.NEW && this.status != RequestStatus.REVIEWING && this.status != RequestStatus.ASSIGNED) {
            throw new DomainException("Cannot cancel a request in '" + this.status + "' status. Only requests in 'NEW', 'REVIEWING', or 'ASSIGNED' status can be cancelled.");
        }
        transitionTo(RequestStatus.CANCELLED, changedByUserId, utcNow, null);
    }

    // ──────────────────────────────────────────────
    // Private Helpers
    // ──────────────────────────────────────────────

    private void transitionTo(RequestStatus newStatus, UUID changedByUserId, Instant utcNow, String note) {
        StatusHistoryEntry entry = new StatusHistoryEntry(null, this, this.status, newStatus, changedByUserId, utcNow, note);
        this.statusHistory.add(entry);
        this.status = newStatus;
        this.updatedAt = utcNow != null ? utcNow : Instant.now();
    }

    private void ensureValidTransition(RequestStatus expectedCurrent, RequestStatus target) {
        if (this.status != expectedCurrent) {
            throw new DomainException("Cannot transition to '" + target + "' from '" + this.status + "'. Expected current status: '" + expectedCurrent + "'.");
        }
    }

    private void ensureNotInTerminalState(String action) {
        if (this.status == RequestStatus.CLOSED || this.status == RequestStatus.REJECTED || this.status == RequestStatus.CANCELLED) {
            throw new DomainException("Cannot " + action + " on a request in terminal state '" + this.status + "'.");
        }
    }

    // Getters
    public UUID getId() { return id; }
    public String getTitle() { return title; }
    public String getDescription() { return description; }
    public UUID getCategoryId() { return categoryId; }
    public UUID getCitizenId() { return citizenId; }
    public RequestStatus getStatus() { return status; }
    public Priority getPriority() { return priority; }
    public UUID getAssignedDepartmentId() { return assignedDepartmentId; }
    public UUID getAssignedEmployeeId() { return assignedEmployeeId; }
    public Location getLocation() { return location; }
    public Instant getCreatedAt() { return createdAt; }
    public Instant getUpdatedAt() { return updatedAt; }
    public List<StatusHistoryEntry> getStatusHistory() { return Collections.unmodifiableList(statusHistory); }
    public List<Comment> getComments() { return Collections.unmodifiableList(comments); }
    public List<Attachment> getAttachments() { return Collections.unmodifiableList(attachments); }
}

File: backend/dotnet/src/BCSMS.Domain/Entities/ServiceRequest.cs
using BCSMS.Domain.Common;
using BCSMS.Domain.Enums;
using BCSMS.Domain.ValueObjects;

namespace BCSMS.Domain.Entities;

/// <summary>
/// Represents a citizen's request for a municipal service.
/// Aggregate root containing status history, comments, and attachment metadata.
/// Encapsulates the full request lifecycle state machine.
///
/// All timestamps are supplied by the caller to keep domain behavior deterministic
/// and unit-testable.
/// </summary>
public class ServiceRequest : BaseEntity
{
    public string Title { get; private set; } = default!;
    public string? Description { get; private set; }
    public Guid CategoryId { get; private set; }
    public RequestStatus Status { get; private set; }
    public Priority? Priority { get; private set; }
    public Location? Location { get; private set; }

    /// <summary>
    /// The citizen who created this request. Immutable after creation.
    /// </summary>
    public Guid CitizenId { get; private set; }

    public Guid? AssignedDepartmentId { get; private set; }
    public Guid? AssignedEmployeeId { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime? UpdatedAt { get; private set; }

    // Encapsulated collections — external code uses the IReadOnlyList properties.
    private readonly List<StatusHistoryEntry> _statusHistory = new();
    private readonly List<Comment> _comments = new();
    private readonly List<Attachment> _attachments = new();

    public IReadOnlyList<StatusHistoryEntry> StatusHistory => _statusHistory.AsReadOnly();
    public IReadOnlyList<Comment> Comments => _comments.AsReadOnly();
    public IReadOnlyList<Attachment> Attachments => _attachments.AsReadOnly();

    private ServiceRequest() : base()
    {
        // For EF Core
    }

    public ServiceRequest(
        Guid id,
        string title,
        Guid categoryId,
        Guid citizenId,
        DateTime createdAt,
        string? description = null,
        Location? location = null)
        : base(id)
    {
        if (string.IsNullOrWhiteSpace(title))
            throw new DomainException("Service request title is required.");

        if (categoryId == Guid.Empty)
            throw new DomainException("Category ID is required.");

        if (citizenId == Guid.Empty)
            throw new DomainException("Citizen ID is required.");

        Title = title.Trim();
        Description = description?.Trim();
        CategoryId = categoryId;
        CitizenId = citizenId;
        Location = location;
        Status = RequestStatus.New;
        CreatedAt = createdAt;
    }

    // ──────────────────────────────────────────────
    // State Machine Transitions
    // ──────────────────────────────────────────────

    /// <summary>
    /// Transitions: New → Reviewing.
    /// A manager or admin begins reviewing the request.
    /// </summary>
    public void StartReview(Guid changedByUserId, DateTime utcNow)
    {
        EnsureValidTransition(RequestStatus.New, RequestStatus.Reviewing);
        TransitionTo(RequestStatus.Reviewing, changedByUserId, utcNow);
    }

    /// <summary>
    /// Transitions: Reviewing → Assigned.
    /// Assigns the request to a department and optionally to an employee.
    /// Priority must be provided.
    /// </summary>
    public void Assign(Guid departmentId, Guid? employeeId, Priority priority,
        Guid changedByUserId, DateTime utcNow)
    {
        EnsureValidTransition(RequestStatus.Reviewing, RequestStatus.Assigned);

        if (departmentId == Guid.Empty)
            throw new DomainException("Department ID is required for assignment.");

        AssignedDepartmentId = departmentId;
        AssignedEmployeeId = employeeId;
        Priority = priority;
        TransitionTo(RequestStatus.Assigned, changedByUserId, utcNow);
    }

    /// <summary>
    /// Transitions: Assigned → InProgress.
    /// The assigned employee begins working on the request.
    /// </summary>
    public void StartProgress(Guid changedByUserId, DateTime utcNow)
    {
        EnsureValidTransition(RequestStatus.Assigned, RequestStatus.InProgress);
        TransitionTo(RequestStatus.InProgress, changedByUserId, utcNow);
    }

    /// <summary>
    /// Transitions: InProgress → Resolved.
    /// The employee has completed the work.
    /// </summary>
    public void Resolve(Guid changedByUserId, DateTime utcNow, string? note = null)
    {
        EnsureValidTransition(RequestStatus.InProgress, RequestStatus.Resolved);
        TransitionTo(RequestStatus.Resolved, changedByUserId, utcNow, note);
    }

    /// <summary>
    /// Transitions: Resolved → Closed.
    /// A manager or admin confirms the resolution.
    /// </summary>
    public void Close(Guid changedByUserId, DateTime utcNow, string? note = null)
    {
        EnsureValidTransition(RequestStatus.Resolved, RequestStatus.Closed);
        TransitionTo(RequestStatus.Closed, changedByUserId, utcNow, note);
    }

    /// <summary>
    /// Transitions: Resolved → InProgress.
    /// Resolution was not satisfactory — reopen for further work.
    /// </summary>
    public void Reopen(Guid changedByUserId, DateTime utcNow, string? note = null)
    {
        EnsureValidTransition(RequestStatus.Resolved, RequestStatus.InProgress);
        TransitionTo(RequestStatus.InProgress, changedByUserId, utcNow, note);
    }

    /// <summary>
    /// Transitions: New or Reviewing → Rejected.
    /// The request is denied by staff.
    /// </summary>
    public void Reject(Guid changedByUserId, DateTime utcNow, string? note = null)
    {
        if (Status is not (RequestStatus.New or RequestStatus.Reviewing))
            throw new DomainException(
                $"Cannot reject a request in '{Status}' status. " +
                "Only 'New' or 'Reviewing' requests can be rejected.");

        TransitionTo(RequestStatus.Rejected, changedByUserId, utcNow, note);
    }

    /// <summary>
    /// Transitions: New, Reviewing, or Assigned → Cancelled.
    /// The citizen withdraws the request, or a manager cancels it before work starts.
    /// </summary>
    public void Cancel(Guid changedByUserId, DateTime utcNow)
    {
        if (Status is not (RequestStatus.New or RequestStatus.Reviewing or RequestStatus.Assigned))
            throw new DomainException(
                $"Cannot cancel a request in '{Status}' status. " +
                "Only requests in 'New', 'Reviewing', or 'Assigned' status can be cancelled.");

        TransitionTo(RequestStatus.Cancelled, changedByUserId, utcNow);
    }

    // ──────────────────────────────────────────────
    // Child Entity Management
    // ──────────────────────────────────────────────

    /// <summary>
    /// Adds a comment to the request. Not allowed on terminal-state requests.
    /// </summary>
    public Comment AddComment(string content, Guid createdByUserId, DateTime utcNow)
    {
        EnsureNotInTerminalState("add a comment");

        var comment = new Comment(Guid.NewGuid(), Id, content, createdByUserId, utcNow);
        _comments.Add(comment);
        UpdatedAt = utcNow;
        return comment;
    }

    /// <summary>
    /// Adds attachment metadata to the request. Not allowed on terminal-state requests.
    /// </summary>
    public Attachment AddAttachment(
        string fileName,
        string contentType,
        long fileSizeInBytes,
        string storagePath,
        Guid uploadedByUserId,
        DateTime utcNow)
    {
        EnsureNotInTerminalState("add an attachment");

        var attachment = new Attachment(
            Guid.NewGuid(), Id, fileName, contentType,
            fileSizeInBytes, storagePath, uploadedByUserId, utcNow);
        _attachments.Add(attachment);
        UpdatedAt = utcNow;
        return attachment;
    }

    // ──────────────────────────────────────────────
    // Detail Update Methods
    // ──────────────────────────────────────────────

    /// <summary>
    /// Updates the request title. Not allowed on terminal-state requests.
    /// </summary>
    public void UpdateTitle(string title, DateTime utcNow)
    {
        EnsureNotInTerminalState("update title");

        if (string.IsNullOrWhiteSpace(title))
            throw new DomainException("Service request title cannot be empty.");

        Title = title.Trim();
        UpdatedAt = utcNow;
    }

    /// <summary>
    /// Updates the request description. Not allowed on terminal-state requests.
    /// </summary>
    public void UpdateDescription(string? description, DateTime utcNow)
    {
        EnsureNotInTerminalState("update description");

        Description = description?.Trim();
        UpdatedAt = utcNow;
    }

    /// <summary>
    /// Changes the request category. Not allowed on terminal-state requests.
    /// </summary>
    public void ChangeCategory(Guid categoryId, DateTime utcNow)
    {
        EnsureNotInTerminalState("change category");

        if (categoryId == Guid.Empty)
            throw new DomainException("Category ID is required.");

        CategoryId = categoryId;
        UpdatedAt = utcNow;
    }

    /// <summary>
    /// Sets or changes the request location. Not allowed on terminal-state requests.
    /// </summary>
    public void ChangeLocation(Location location, DateTime utcNow)
    {
        EnsureNotInTerminalState("change location");

        Location = location ?? throw new DomainException("Location is required. Use RemoveLocation to clear it.");
        UpdatedAt = utcNow;
    }

    /// <summary>
    /// Removes the request location. Not allowed on terminal-state requests.
    /// </summary>
    public void RemoveLocation(DateTime utcNow)
    {
        EnsureNotInTerminalState("remove location");

        Location = null;
        UpdatedAt = utcNow;
    }

    /// <summary>
    /// Sets or updates the priority. Not allowed on terminal-state requests.
    /// </summary>
    public void SetPriority(Priority priority, DateTime utcNow)
    {
        EnsureNotInTerminalState("set priority");

        Priority = priority;
        UpdatedAt = utcNow;
    }

    // ──────────────────────────────────────────────
    // Private Helpers
    // ──────────────────────────────────────────────

    private void TransitionTo(RequestStatus newStatus, Guid changedByUserId,
        DateTime utcNow, string? note = null)
    {
        var entry = new StatusHistoryEntry(
            Guid.NewGuid(), Id, Status, newStatus, changedByUserId, utcNow, note);

        _statusHistory.Add(entry);
        Status = newStatus;
        UpdatedAt = utcNow;
    }

    private void EnsureValidTransition(RequestStatus expectedCurrent, RequestStatus target)
    {
        if (Status != expectedCurrent)
            throw new DomainException(
                $"Cannot transition to '{target}' from '{Status}'. " +
                $"Expected current status: '{expectedCurrent}'.");
    }

    private void EnsureNotInTerminalState(string action)
    {
        if (Status is RequestStatus.Closed or RequestStatus.Rejected or RequestStatus.Cancelled)
            throw new DomainException(
                $"Cannot {action} on a request in terminal state '{Status}'.");
    }
}
